File: world/tests/test_world.py
"""tests/test_world.py - Basic smoke tests."""
import numpy as np
import world
from world.spaces import Discrete, Box
from world.wrappers import TimeLimitWrapper, RecordEpisodeWrapper


# ─────────────────────────────────────────────────────────────────────────────
#  Space tests
# ─────────────────────────────────────────────────────────────────────────────

def test_discrete_space():
    sp = Discrete(4, seed=0)
    for _ in range(50):
        s = sp.sample()
        assert sp.contains(s), f"{s} not in {sp}"
    assert sp.contains(0)
    assert sp.contains(3)
    assert not sp.contains(4)
    assert not sp.contains(-1)
    print("  [PASS] Discrete space")


def test_box_space():
    sp = Box(-1.0, 1.0, shape=(3,), seed=0)
    for _ in range(50):
        s = sp.sample()
        assert sp.contains(s), f"{s} not in {sp}"
    assert sp.contains(np.array([0.0, 0.5, -0.9], dtype=np.float32))
    assert not sp.contains(np.array([1.1, 0.0, 0.0], dtype=np.float32))
    print("  [PASS] Box space")


# ─────────────────────────────────────────────────────────────────────────────
#  FrozenLake tests
# ─────────────────────────────────────────────────────────────────────────────

def test_frozen_lake_reset():
    env = world.make("FrozenLake-v1")
    obs, info = env.reset(seed=0)
    assert obs == 0, f"Expected start state 0, got {obs}"
    assert "pos" in info
    print("  [PASS] FrozenLake reset")


def test_frozen_lake_step():
    env = world.make("FrozenLake-v0")   # deterministic
    obs, _ = env.reset(seed=1)
    result = env.step(2)   # RIGHT
    assert isinstance(result.observation, (int, np.integer))
    assert isinstance(result.reward, float)
    assert isinstance(result.terminated, bool)
    assert isinstance(result.truncated, bool)
    print("  [PASS] FrozenLake step")


def test_frozen_lake_unpack():
    env = world.make("FrozenLake-v0")
    env.reset()
    obs, reward, terminated, truncated, info = env.step(1)
    assert isinstance(obs, (int, np.integer))
    print("  [PASS] FrozenLake tuple unpack")


def test_frozen_lake_full_episode():
    env = world.make("FrozenLake-v0")
    obs, _ = env.reset(seed=42)
    done = False
    steps = 0
    while not done and steps < 200:
        result = env.step(env.action_space.sample())
        obs = result.observation
        done = result.done
        steps += 1
    env.close()
    print(f"  [PASS] FrozenLake full episode ({steps} steps)")


# ─────────────────────────────────────────────────────────────────────────────
#  CartPole tests
# ─────────────────────────────────────────────────────────────────────────────

def test_cartpole_reset():
    env = world.make("CartPole-v1")
    obs, info = env.reset(seed=0)
    assert obs.shape == (4,), f"Expected shape (4,), got {obs.shape}"
    assert env.observation_space.contains(obs)
    print("  [PASS] CartPole reset")


def test_cartpole_step():
    env = world.make("CartPole-v1")
    env.reset(seed=0)
    result = env.step(0)
    assert result.observation.shape == (4,)
    assert result.reward == 1.0
    print("  [PASS] CartPole step")


def test_cartpole_full_episode():
    env = world.make("CartPole-v1")
    obs, _ = env.reset(seed=7)
    total_reward = 0.0
    done = False
    while not done:
        result = env.step(env.action_space.sample())
        total_reward += result.reward
        done = result.done
    env.close()
    assert total_reward >= 1.0
    print(f"  [PASS] CartPole full episode (reward={total_reward:.0f})")


def test_cartpole_invalid_action():
    env = world.make("CartPole-v1")
    env.reset()
    try:
        env.step(5)
        assert False, "Should have raised ValueError"
    except ValueError:
        pass
    print("  [PASS] CartPole invalid action raises ValueError")


# ─────────────────────────────────────────────────────────────────────────────
#  Wrapper tests
# ─────────────────────────────────────────────────────────────────────────────

def test_time_limit_wrapper():
    env = TimeLimitWrapper(world.make("CartPole-v1"), max_steps=5)
    env.reset(seed=0)
    for i in range(4):
        result = env.step(0)
        assert not result.truncated or result.terminated
    result = env.step(0)
    assert result.truncated or result.terminated
    env.close()
    print("  [PASS] TimeLimitWrapper")


def test_record_episode_wrapper():
    env = RecordEpisodeWrapper(world.make("FrozenLake-v0"))
    for ep in range(5):
        env.reset(seed=ep)
        done = False
        while not done:
            result = env.step(env.action_space.sample())
            done = result.done
    assert len(env.episode_stats) == 5
    s = env.summary()
    assert "mean_reward" in s
    env.close()
    print(f"  [PASS] RecordEpisodeWrapper (mean_reward={s['mean_reward']:.2f})")


# ─────────────────────────────────────────────────────────────────────────────
#  Registry tests
# ─────────────────────────────────────────────────────────────────────────────

def test_registry():
    r = world.registry()
    for key in ["FrozenLake-v0", "FrozenLake-v1", "FrozenLake8x8-v1", "CartPole-v1"]:
        assert key in r, f"{key} not in registry"
    print("  [PASS] Registry contains all built-in envs")


def test_make_unknown():
    try:
        world.make("NoSuchEnv-v999")
        assert False
    except KeyError:
        pass
    print("  [PASS] make() raises KeyError for unknown env")


# ─────────────────────────────────────────────────────────────────────────────
#  Run all
# ─────────────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    tests = [
        test_discrete_space,
        test_box_space,
        test_frozen_lake_reset,
        test_frozen_lake_step,
        test_frozen_lake_unpack,
        test_frozen_lake_full_episode,
        test_cartpole_reset,
        test_cartpole_step,
        test_cartpole_full_episode,
        test_cartpole_invalid_action,
        test_time_limit_wrapper,
        test_record_episode_wrapper,
        test_registry,
        test_make_unknown,
    ]
    print(f"\n{'='*50}")
    print(f"  world test suite  ({len(tests)} tests)")
    print(f"{'='*50}")
    passed = 0
    for t in tests:
        try:
            t()
            passed += 1
        except Exception as e:
            print(f"  [FAIL] {t.__name__}: {e}")
    print(f"{'─'*50}")
    print(f"  {passed}/{len(tests)} passed")
    print(f"{'='*50}\n")

File: world/tests/test_world.ts
/**
 * tests/test_world.ts - Basic smoke tests for world package.
 */
import { Discrete, Box } from '../spaces';
import { Env, StepResult } from '../core';
import { TimeLimitWrapper, RecordEpisodeWrapper } from '../wrappers';
import { make, registry } from '../utils/registry';

function assert(condition: boolean, msg: string) {
  if (!condition) throw new Error(`Assertion failed: ${msg}`);
}

function assertAlmostEq(a: number, b: number, tol = 1e-5) {
  if (Math.abs(a - b) > tol) throw new Error(`Expected ${a} ≈ ${b}`);
}

// ─────────────────────────────────────────────────────────────────────────────
// Space tests
// ─────────────────────────────────────────────────────────────────────────────

function testDiscreteSpace() {
  const sp = new Discrete(4);
  for (let i = 0; i < 50; i++) {
    const s = sp.sample();
    assert(sp.contains(s), `${s} not in Discrete(4)`);
  }
  assert(sp.contains(0), 'should contain 0');
  assert(sp.contains(3), 'should contain 3');
  assert(!sp.contains(4), 'should not contain 4');
  assert(!sp.contains(-1), 'should not contain -1');
  console.log('  [PASS] Discrete space');
}

function testBoxSpace() {
  const sp = new Box(-1.0, 1.0, [3]);
  for (let i = 0; i < 50; i++) {
    const s = sp.sample();
    assert(sp.contains(s), `${s} not in Box(-1,1,shape=3)`);
  }
  const inside = [0.0, 0.5, -0.9];
  assert(sp.contains(inside), 'should contain [0, 0.5, -0.9]');
  const outside = [1.1, 0.0, 0.0];
  assert(!sp.contains(outside), 'should not contain [1.1, 0, 0]');
  console.log('  [PASS] Box space');
}

// ─────────────────────────────────────────────────────────────────────────────
// FrozenLake tests
// ─────────────────────────────────────────────────────────────────────────────

function testFrozenLakeReset() {
  const env = make('FrozenLake-v1') as Env<number, number>;
  const result = env.reset({ seed: 0 });
  assert(result.observation === 0, `Expected start state 0, got ${result.observation}`);
  assert('pos' in result.info, 'info should have pos');
  console.log('  [PASS] FrozenLake reset');
}

function testFrozenLakeStep() {
  const env = make('FrozenLake-v0') as Env<number, number>;
  const result = env.reset({ seed: 1 });
  const stepResult = env.step(2); // RIGHT
  assert(typeof stepResult.observation === 'number', 'observation should be number');
  assert(typeof stepResult.reward === 'number', 'reward should be number');
  assert(typeof stepResult.terminated === 'boolean', 'terminated should be boolean');
  assert(typeof stepResult.truncated === 'boolean', 'truncated should be boolean');
  console.log('  [PASS] FrozenLake step');
}

function testFrozenLakeUnpack() {
  const env = make('FrozenLake-v0') as Env<number, number>;
  env.reset();
  const result = env.step(1);
  assert(typeof result.observation === 'number', 'observation should be number');
  console.log('  [PASS] FrozenLake tuple unpack');
}

function testFrozenLakeFullEpisode() {
  const env = make('FrozenLake-v0') as Env<number, number>;
  env.reset({ seed: 42 });
  let done = false;
  let steps = 0;
  const maxSteps = 200;
  while (!done && steps < maxSteps) {
    const result = env.step(env.actionSpace.sample() as number);
    done = result.done;
    steps++;
  }
  env.close();
  console.log(`  [PASS] FrozenLake full episode (${steps} steps)`);
}

// ─────────────────────────────────────────────────────────────────────────────
// CartPole tests
// ─────────────────────────────────────────────────────────────────────────────

function testCartPoleReset() {
  const env = make('CartPole-v1') as Env<number[], number>;
  const result = env.reset({ seed: 0 });
  assert(result.observation.length === 4, `Expected shape (4,), got (${result.observation.length},)`);
  console.log('  [PASS] CartPole reset');
}

function testCartPoleStep() {
  const env = make('CartPole-v1') as Env<number[], number>;
  env.reset({ seed: 0 });
  const result = env.step(0);
  assert(result.observation.length === 4, `Expected shape (4,), got (${result.observation.length},)`);
  assertAlmostEq(result.reward, 1.0);
  console.log('  [PASS] CartPole step');
}

function testCartPoleFullEpisode() {
  const env = make('CartPole-v1') as Env<number[], number>;
  env.reset({ seed: 7 });
  let totalReward = 0.0;
  let done = false;
  while (!done) {
    const result = env.step(env.actionSpace.sample() as number);
    totalReward += result.reward;
    done = result.done;
  }
  env.close();
  assert(totalReward >= 1.0, `Expected total_reward >= 1, got ${totalReward}`);
  console.log(`  [PASS] CartPole full episode (reward=${totalReward.toFixed(0)})`);
}

function testCartPoleInvalidAction() {
  const env = make('CartPole-v1') as Env<number[], number>;
  env.reset();
  try {
    env.step(5);
    assert(false, 'Should have raised Error');
  } catch (e) {
    // Expected
  }
  console.log('  [PASS] CartPole invalid action raises Error');
}

// ─────────────────────────────────────────────────────────────────────────────
// Wrapper tests
// ─────────────────────────────────────────────────────────────────────────────

function testTimeLimitWrapper() {
  const env = make('CartPole-v1') as Env<number[], number>;
  const wrapped = new TimeLimitWrapper(env, 5);
  wrapped.reset({ seed: 0 });
  for (let i = 0; i < 4; i++) {
    const result = wrapped.step(0);
    assert(!result.truncated || result.terminated, `step ${i} should not truncate`);
  }
  const result = wrapped.step(0);
  assert(result.truncated || result.terminated, 'step 5 should truncate or terminate');
  env.close();
  console.log('  [PASS] TimeLimitWrapper');
}

function testRecordEpisodeWrapper() {
  const env = make('FrozenLake-v0') as Env<number, number>;
  const wrapped = new RecordEpisodeWrapper(env);
  for (let ep = 0; ep < 5; ep++) {
    wrapped.reset({ seed: ep });
    let done = false;
    while (!done) {
      const result = wrapped.step(wrapped.actionSpace.sample() as number);
      done = result.done;
    }
  }
  assert(wrapped.episode_stats.length === 5, `Expected 5 episodes, got ${wrapped.episode_stats.length}`);
  const s = wrapped.summary();
  assert('mean_reward' in s, 'summary should have mean_reward');
  env.close();
  console.log(`  [PASS] RecordEpisodeWrapper (mean_reward=${s.mean_reward.toFixed(2)})`);
}

// ─────────────────────────────────────────────────────────────────────────────
// Registry tests
// ─────────────────────────────────────────────────────────────────────────────

function testRegistry() {
  const r = registry();
  for (const key of ['FrozenLake-v0', 'FrozenLake-v1', 'FrozenLake8x8-v1', 'CartPole-v1']) {
    assert(key in r, `${key} not in registry`);
  }
  console.log('  [PASS] Registry contains all built-in envs');
}

function testMakeUnknown() {
  try {
    make('NoSuchEnv-v999');
    assert(false, 'Should have thrown');
  } catch (e) {
    // Expected
  }
  console.log('  [PASS] make() raises Error for unknown env');
}

// ─────────────────────────────────────────────────────────────────────────────
// Run all
// ─────────────────────────────────────────────────────────────────────────────

const tests = [
  testDiscreteSpace,
  testBoxSpace,
  testFrozenLakeReset,
  testFrozenLakeStep,
  testFrozenLakeUnpack,
  testFrozenLakeFullEpisode,
  testCartPoleReset,
  testCartPoleStep,
  testCartPoleFullEpisode,
  testCartPoleInvalidAction,
  testTimeLimitWrapper,
  testRecordEpisodeWrapper,
  testRegistry,
  testMakeUnknown,
];

console.log('\n' + '='.repeat(50));
console.log(`  world test suite  (${tests.length} tests)`);
console.log('='.repeat(50));

let passed = 0;
for (const t of tests) {
  try {
    t();
    passed++;
  } catch (e: any) {
    console.log(`  [FAIL] ${t.name}: ${e.message}`);
  }
}
console.log('-'.repeat(50));
console.log(`  ${passed}/${tests.length} passed`);
console.log('='.repeat(50) + '\n');